executor/src/witgen/jit/solver/mod.rs

```diff
@@ -0,0 +1,212 @@
+use powdr_number::FieldElement;
+use utils::known_variables;
+
+use super::effect::Effect;
+use super::quadratic_symbolic_expression::{Error, RangeConstraintProvider};
+use super::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    symbolic_expression::SymbolicExpression,
+};
+use crate::witgen::jit::variable_update::VariableUpdate;
+use crate::witgen::range_constraints::RangeConstraint;
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+#[cfg(test)]
+mod tests;
+mod utils;
+
+/// Current state of a variable in the solver.
+#[derive(Debug, PartialEq)]
+struct VariableState<T: FieldElement, V> {
+    /// Range constraint for the variable.
+    range_constraint: RangeConstraint<T>,
+    /// None if unknown, otherwise a symbolic expression that can be used to compute
+    /// the variable from other variables.
+    /// Note that the expression is an instance of `SymbolicExpression::Concrete(_)`,
+    /// we have found a concrete variable assignment.
+    symbolic_expression: Option<SymbolicExpression<T, V>>,
+}
+
+impl<T: FieldElement, V> Default for VariableState<T, V> {
+    fn default() -> Self {
```

Comment:
Is that not derivable?

Reply:
Yes, but we'd have to require `V: Default` (which is BS, because it only appears in `Option<SymbolicExpression<T, V>>`...). I added a comment.